external-crates/move/move-compiler/src/editions/mod.rs

```diff
@@ -0,0 +1,269 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//! This module controls feature gating and breaking changes in new editions of the source language
+
+use std::{fmt::Display, str::FromStr};
+
+use crate::{diag, shared::CompilationEnv};
+use move_ir_types::location::*;
+use serde::{Deserialize, Serialize};
+
+//**************************************************************************************************
+// types
+//**************************************************************************************************
+
+#[derive(PartialEq, Eq, Clone, Copy, Debug, PartialOrd, Ord)]
+pub enum Edition {
+    Legacy,
+    E2024(EditionRelease),
+}
+
+#[derive(PartialEq, Eq, Clone, Copy, Debug, PartialOrd, Ord)]
+pub enum EditionRelease {
+    Final,
+    Beta,
+    Alpha,
+}
+
+#[derive(PartialEq, Eq, Clone, Copy, Debug)]
+pub enum FeatureGate {
+    LetMut,
+    ReceiverSyntax,
+    MacroFun,
+    UnderscoreType,
+}
+
+#[derive(PartialEq, Eq, Clone, Copy, Debug, PartialOrd, Ord)]
+pub enum Flavor {
+    GlobalStorage,
+    Sui,
+}
+
+//**************************************************************************************************
+// Entry
+//**************************************************************************************************
+
+pub fn check_feature(env: &mut CompilationEnv, edition: Edition, loc: Loc, feature: FeatureGate) {
+    let min_edition = feature.min_edition();
+    if edition < min_edition {
+        env.add_diag(diag!(
+            Editions::FeatureTooNew,
+            (loc, "{} requires edition {} or newer")
+        ))
+    }
+}
+
+//**************************************************************************************************
+// impls
+//**************************************************************************************************
+
+impl Edition {
+    pub const LEGACY: &str = "legacy";
+    pub const E2024_PREFIX: &str = "2024";
+
+    pub const SUPPORTED: &[Self] = &[Self::Legacy, Self::E2024(EditionRelease::Alpha)];
+}
+
+impl EditionRelease {
+    pub const ALPHA: &str = "alpha";
+    pub const BETA: &str = "beta";
+    pub const EXT_SEP: &str = ".";
+
+    pub const SUFFIXES: &[Self] = &[Self::Alpha, Self::Beta];
+}
+
+impl FeatureGate {
+    pub fn min_edition(&self) -> Edition {
+        match self {
+            FeatureGate::LetMut => Edition::E2024(EditionRelease::Alpha),
+            FeatureGate::ReceiverSyntax => Edition::E2024(EditionRelease::Alpha),
+            FeatureGate::MacroFun => Edition::E2024(EditionRelease::Alpha),
+            FeatureGate::UnderscoreType => Edition::E2024(EditionRelease::Alpha),
+        }
+    }
+}
```

Comment:
This may just be me, but it feels more intuitive for me to ask the `Edition` whether it supports a given `Feature(Gate)` than to ask the feature what is the minimal version that supports it -- how would we deal with a situation where a feature starts off supported and then support is removed, otherwise?